DependencyInjection/Compiler/CustomHandlersPass.php

```diff
@@ -61,22 +66,63 @@ public function process(ContainerBuilder $container)
 
                 foreach ($directions as $direction) {
                     $method = isset($methodData['method']) ? $methodData['method'] : HandlerRegistry::getDefaultMethod($direction, $methodData['type'], $methodData['format']);
-                    if (class_exists(ServiceLocatorTagPass::class) || $container->getDefinition($id)->isPublic()) {
-                        $handlerServices[$id] = new Reference($id);
+                    if (class_exists(ServiceLocatorTagPass::class) || $definition->isPublic()) {
+                        $handlerServices[$id] = $reference;
                         $handlers[$direction][$methodData['type']][$methodData['format']] = array($id, $method);
                     } else {
-                        $handlers[$direction][$methodData['type']][$methodData['format']] = array(new Reference($id), $method);
+                        $handlers[$direction][$methodData['type']][$methodData['format']] = array($reference, $method);
                     }
                 }
             }
         }
 
-        $container->findDefinition('jms_serializer.handler_registry')
-            ->addArgument($handlers);
+        $container->findDefinition('jms_serializer.handler_registry')->addArgument($handlers);
 
         if (class_exists(ServiceLocatorTagPass::class)) {
             $serviceLocator = ServiceLocatorTagPass::register($container, $handlerServices);
             $container->findDefinition('jms_serializer.handler_registry')->replaceArgument(0, $serviceLocator);
         }
     }
+
+    /**
+     * Finds all services with the given tag name and order them by their priority.
+     *
+     * The order of additions must be respected for services having the same priority,
+     * and knowing that the \SplPriorityQueue class does not respect the FIFO method,
+     * we should not use that class.
+     *
+     * Original author is Iltar van der Berg <kjarli@gmail.com>
+     *
+     * @see https://bugs.php.net/bug.php?id=53710
+     * @see https://bugs.php.net/bug.php?id=60926
+     *
+     * @param string           $tagName
+     * @param ContainerBuilder $container
+     *
+     * @return Reference[]
+     */
+    private function findAndSortTaggedServices($tagName, ContainerBuilder $container)
+    {
+        $services = array();
+
+        foreach ($container->findTaggedServiceIds($tagName, true) as $serviceId => $attributes) {
+            $definition = $container->getDefinition($serviceId);
+
+            if (!isset($attributes[0]['priority'])) {
```

Comment:
prefer to keep this simpler without special cases. if the priority is not set, just set it to `0`. users can still control priorities by adding their services by using negative priorities